core/src/main/scala/cats/FunctorFilter.scala

```diff
@@ -0,0 +1,62 @@
+package cats
+
+import simulacrum.typeclass
+
+@typeclass trait FunctorFilter[F[_]] extends Functor[F] {
+
+  /**
+   * A combined [[map]] and [[filter]]. Filtering is handled via `Option`
+   * instead of `Boolean` such that the output type `B` can be different than
+   * the input type `A`.
+   *
+   * Example:
+   * {{{
+   * scala> import cats.implicits._
+   * scala> val m: Map[Int, String] = Map(1 -> "one", 3 -> "three")
+   * scala> val l: List[Int] = List(1, 2, 3, 4)
+   * scala> def asString(i: Int): Option[String] = m.get(i)
+   * scala> l.mapFilter(i => m.get(i))
+   * res0: List[String] = List(one, three)
+   * }}}
+   */
+  def mapFilter[A, B](fa: F[A])(f: A => Option[B]): F[B]
```

Comment:
Actually, when you think about it, it might be something like this:
```scala
trait FunctorArrow[F[_], A[_, _]] extends Functor[F] {
  def arrow: Arrow[A]
  def mapArrow[T, U](f: F[T], arrow: A[T, U]): F[U]
}
```
And we are just talking about various `KleisliArrows` here.

Or maybe we mean something like:

```scala
trait FunctorTraverse[F[_], M[_]] extends Functor[F] {
  // this is similar to traverse, except we end up with `F[B]` not `M[F[B]]`
  def mapTraverse[A, B](fa: F[A])(f: A => M[B]): F[B]
}
```
It is not clear what the right level of generalization is. It seems to me `A => M[B]` where you have `Foldable[M]` is pretty generally useful. Maybe that is the thing you want instead of mapConcat:
```scala
trait FunctorFoldableFn[F[_]] extend Functor[F] {
  def mapFoldable[A, B, M[_]](f: F[A])(f: A => M[B])(implicit M: Foldable[M]): F[A]
}
```